modules/rgbd/src/hash_tsdf.cpp

```diff
@@ -0,0 +1,661 @@
+    if (it == volumeUnits.end())
+    {
+        TsdfVoxel dummy;
+        dummy.tsdf   = 1.f;
+        dummy.weight = 0;
+        return dummy;
+    }
+    cv::Ptr<TSDFVolumeCPU> volumeUnit =
+        std::dynamic_pointer_cast<TSDFVolumeCPU>(it->second.pVolume);
+
+    cv::Point3f volumeUnitPos = volumeUnitIdxToVolume(volumeUnitIdx);
+    cv::Vec3i volUnitLocalIdx = volumeToVoxelCoord(point - volumeUnitPos);
+    volUnitLocalIdx =
+        cv::Vec3i(abs(volUnitLocalIdx[0]), abs(volUnitLocalIdx[1]), abs(volUnitLocalIdx[2]));
+    return volumeUnit->at(volUnitLocalIdx);
+}
+
+inline Point3f HashTSDFVolumeCPU::getNormalVoxel(Point3f point) const
+{
+    Vec3f pointVec(point);
+    Vec3f normal = Vec3f(0, 0, 0);
+
+    Vec3f pointPrev = point;
+    Vec3f pointNext = point;
+
+    for (int c = 0; c < 3; c++)
+    {
+        pointPrev[c] -= voxelSize * 0.5;
+        pointNext[c] += voxelSize * 0.5;
+
+        normal[c] = at(Point3f(pointNext)).tsdf - at(Point3f(pointPrev)).tsdf;
+        normal[c] *= 0.5;
+
+        pointPrev[c] = pointVec[c];
+        pointNext[c] = pointVec[c];
+    }
+    return normalize(normal);
+}
+
+struct RaycastInvoker : ParallelLoopBody
+{
+    RaycastInvoker(Points& _points, Normals& _normals, Affine3f cameraPose, Intr intrinsics,
+                   const HashTSDFVolumeCPU& _volume)
+        : ParallelLoopBody(),
+          points(_points),
+          normals(_normals),
+          volume(_volume),
+          tstep(_volume.truncDist * _volume.raycastStepFactor),
+          cam2vol(volume.pose.inv() * cameraPose),
+          vol2cam(cameraPose.inv() * volume.pose),
+          reproj(intrinsics.makeReprojector())
+    {
+    }
+
+    virtual void operator()(const Range& range) const override
+    {
+        const Point3f cam2volTrans = cam2vol.translation();
+        const Matx33f cam2volRot   = cam2vol.rotation();
+        const Matx33f vol2camRot   = vol2cam.rotation();
+
+        const float blockSize = volume.volumeUnitSize;
+
+        for (int y = range.start; y < range.end; y++)
+        {
+            ptype* ptsRow = points[y];
+            ptype* nrmRow = normals[y];
+
+            for (int x = 0; x < points.cols; x++)
+            {
+                //! Initialize default value
+                Point3f point = nan3, normal = nan3;
+
+                //! Ray origin and direction in the volume coordinate frame
+                Point3f orig     = cam2volTrans;
+                Point3f rayDirV  = normalize(Vec3f(cam2volRot * reproj(Point3f(x, y, 1.f))));
+
+                float tmin = 0;
+                float tmax = volume.truncateThreshold / rayDirV.z;
+                float tcurr = tmin;
+
+                cv::Vec3i prevVolumeUnitIdx =
+                    cv::Vec3i(std::numeric_limits<int>::min(), std::numeric_limits<int>::min(),
+                              std::numeric_limits<int>::min());
+
+                float tprev = tcurr;
+                TsdfType prevTsdf = float16_t(volume.truncDist);
```

Comment:
Looks like you did some experiments with `float16` but forgot to delete it from here.